### PR TITLE
Messaging

### PR DESCRIPTION
**Pull request by hadleynet**
Supports all message types in the spec. No doubt will benefit from additional testing and refinements in due course but the happy path is well covered by VRDR.Tests now with some negative testing.

### CODE REVIEW

**Comment by rbclark** [comment]
In order for this to push to Nuget when this is merged, we would need to create the following block in `.github/workflows/publish.yml`

```yaml
      # Publish
      - name: publish on version change
        uses: rohith/publish-nuget@v2
        with:
          PROJECT_FILE_PATH: VRDR.Messaging/VRDRMessaging.csproj # Relative to repository root
          VERSION_FILE_PATH: VRDR.Messaging/VRDRMessaging.csproj # Filepath with version info, relative to repository root. Defaults to project file
          VERSION_REGEX: <Version>(.*)<\/Version> # Regex pattern to extract version info in a capturing group
          TAG_COMMIT: true # Tag the project when the version has changed
          TAG_FORMAT: v* # Format of the git tag, [*] gets replaced with version
          NUGET_KEY: ${{secrets.NUGET_API_KEY}} # nuget.org API key
          PACKAGE_NAME: VRDR.Messaging
```

In order for this to work I believe we will also need to first manually upload a version to Nuget and then do a GitHub release after that. I currently have an outstanding preview version of VRDR called `v3.1.0-preview1` so we can test with `v3.1.0-preview2` if we want to verify the releasing properly works for `VRDR` and `VRDR.Messaging`.

File: VRDR.CLI/Program.cs [hunk 1]
```diff
@@ -620,9 +622,31 @@ static int Main(string[] args)
...
             }
+            else if (args.Length == 3 && args[0] == "sendMessage")
+            {
+                DeathRecord record = new DeathRecord(File.ReadAllText(args[1]));
+
+                for (int i = 0; i < 1000; i++)
+                {
+                    DeathRecordSubmission submission = new DeathRecordSubmission(record);
+                    submission.MessageTimestamp = DateTime.Now;
+                    submission.MessageId = "adam-" + Convert.ToString(i) + "-" + Guid.NewGuid();
```
Comment (holmesie) [inline VRDR.CLI/Program.cs:633]: I think the whole test block can be removed, or at least the `adam-` bit.

File: VRDR.CLI/Program.cs [hunk 2]
```diff
@@ -620,9 +622,29 @@ static int Main(string[] args)
...
+
+                for (int i = 0; i < 1000; i++)
+                {
+                    DeathRecordSubmission submission = new DeathRecordSubmission(record);
+                    StringContent stringContent = new StringContent(submission.ToJSON(), Encoding.UTF8, "application/fhir+json");
+                    CallEndpoint(args[2], stringContent);
+                    Console.WriteLine("Sending " + Convert.ToString(i));
+                }
+
+                System.Threading.Thread.Sleep(10000);
```
[the remark this answers is not in the document]
Comment (holmesie) [reply VRDR.CLI/Program.cs:637]: My comment seems to have disappeared somehow, but I think this whole `sendMessage` block needs to be removed. If we want something like this, it should be done in Canary.

[the remark this answers is not in the document]
Comment (hadleynet) [reply VRDR.CLI/Program.cs:637]: I removed the adam- bit earlier, that probably removed your comment. When you say "sendMessage" block you mean remove this whole operation from the CLI or just drop the 1000.times loop?

File: .github/workflows/run-tests.yml [hunk 3]
```diff
@@ -16,7 +16,7 @@ jobs:
     - name: Setup .NET Core
       uses: actions/setup-dotnet@v1
       with:
-        dotnet-version: 3.1.201
```
[the remark this answers is not in the document]
Comment (hadleynet) [reply .github/workflows/run-tests.yml:19]: I switched the target framework for the libraries to improve compatibility, apparently this line doesn't install the required runtime libraries to run a netcoreapp2.1 app.

File: VRDR.CLI/Program.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (holmesie) [reply VRDR.CLI/Program.cs:637]: The whole block; it was something I whipped up quickly during our hackathon to throw input at what Pete was doing.

File: README.md [hunk 4]
```diff
@@ -159,6 +159,132 @@ CauseCodes example = new CauseCodes(json);
...
+
+```cs
+// Get the DeathRecordSubmission message JSON or XML representation
+StreamReader messageStream = ...;
+
+// Parse the JSON
+DeathRecordSubmission message = (DeathRecordSubmission)BaseMessage.Parse(messageStream);
+
+// Get the DeathRecord
+DeathRecord record = message.MessagePayload;
```
[the remark this answers is not in the document]
Comment (hadleynet) [reply README.md:193]: Will fix

File: README.md [hunk 5]
```diff
@@ -159,6 +159,132 @@ CauseCodes example = new CauseCodes(json);
...
+
+Note that the `AckMessage` constructor will automatically set the message header properties to identify the `DeathRecordSubmission` message that it acknowledges. It will also set the `MessageDestination` property to the value of the `DeathRecordSubmission.MessageSource` property.
+
+#### Creating a Coding Response Message
+
+```cs
+// Create an empty coding response message
+CodingResponseMessage message = new CodingResponseMessage("https://example.org/jurisdiction/endpoint");
+
+// Assign the certificate number and satte identifier
```
[the remark this answers is not in the document]
Comment (hadleynet) [reply README.md:223]: Will fix

File: README.md [hunk 6]
```diff
@@ -159,6 +159,132 @@ CauseCodes example = new CauseCodes(json);
...
+
+var entityAxisEntries = new List<CauseOfDeathEntityAxisEntry>();
+var entry1 = new CauseOfDeathEntityAxisEntry("DEATH CERT LINE 1 TEXT", "ID of VRDR CauseOfDeathCondition");
+entry1.AssignedCodes.Add("...");
+entry1.AssignedCodes.Add("...");
+entityAxisEntries.Add(entry1);
+var entry2 = new CauseOfDeathEntityAxisEntry("DEATH CERT LINE 2 TEXT", "ID of VRDR CauseOfDeathCondition");
+entry2.AssignedCodes.Add("...");
+entityAxisEntries.Add(entry2);
+message.CauseOfDeathEntityAxis = entityAxisEntries;
```
[the remark this answers is not in the document]
Comment (hadleynet) [reply README.md:258]: Maybe support both approaches?

File: README.md [hunk 7]
```diff
@@ -159,6 +159,132 @@ CauseCodes example = new CauseCodes(json);
...
+// Create the coding error message
+var errMsg = CodingErrorMessage(certificateNumber, stateIdentifier, destinationEndpoint);
+
+// Add the issues found during processing
+var issues = new List<Issue>();
+var issue = new Issue(OperationOutcome.IssueSeverity.Fatal, OperationOutcome.IssueType.Invalid, "Description of first issues");
+issues.Add(issue);
+issue = new Issue(OperationOutcome.IssueSeverity.Warning, OperationOutcome.IssueType.Expired, "Description of second issues");
+issues.Add(issue);
+err.Issues = issues;
```
[the remark this answers is not in the document]
Comment (hadleynet) [reply README.md:279]: Support both approaches?

[the remark this answers is not in the document]
Comment (hadleynet) [reply README.md:279]: Will fix

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (pkmitre) [reply README.md:258]: I like one or the other... is there a good reason to require / allow the user of the interface to manage a List themselves as an alternative?

[the remark this answers is not in the document]
Comment (holmesie) [reply README.md:258]: That's how we manage Race and Ethnicity right now.

[the remark this answers is not in the document]
Comment (hadleynet) [reply README.md:258]: The getter returns a list so having a setter that can accept a modified list seems like a win.

[the remark this answers is not in the document]
Comment (pkmitre) [reply README.md:258]: That makes sense; given that, I'm inclined to leave the current interface.

**Pull request merged**